test/e2e/tests/test_user.py

```diff
@@ -0,0 +1,116 @@
+# Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+#	 http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""Integration tests for the IAM User resource"""
+
+import logging
+import time
+
+import pytest
+
+from acktest.k8s import condition
+from acktest.k8s import resource as k8s
+from acktest.resources import random_suffix_name
+from e2e import service_marker, CRD_GROUP, CRD_VERSION, load_resource
+from e2e.common.types import USER_RESOURCE_PLURAL
+from e2e.replacement_values import REPLACEMENT_VALUES
+from e2e import user
+from e2e import tag
+
+DELETE_WAIT_AFTER_SECONDS = 10
+CHECK_STATUS_WAIT_SECONDS = 10
+MODIFY_WAIT_AFTER_SECONDS = 10
+
+
+@service_marker
+@pytest.mark.canary
+class TestUser:
+    def test_crud(self):
+        user_name = random_suffix_name("my-simple-user", 24)
+
+        replacements = REPLACEMENT_VALUES.copy()
+        replacements['USER_NAME'] = user_name
+
+        resource_data = load_resource(
+            "user_simple",
+            additional_replacements=replacements,
+        )
+
+        ref = k8s.CustomResourceReference(
+            CRD_GROUP, CRD_VERSION, USER_RESOURCE_PLURAL,
+            user_name, namespace="default",
+        )
+        k8s.create_custom_resource(ref, resource_data)
+        cr = k8s.wait_resource_consumed_by_controller(ref)
+
+        user.wait_until_exists(user_name)
+
+        time.sleep(CHECK_STATUS_WAIT_SECONDS)
+
+        condition.assert_synced(ref)
+
+        latest = user.get(user_name)
+        assert latest is not None
+        assert latest['UserName'] == user_name
+
+        latest_policy_arns = user.get_attached_policy_arns(user_name)
+        assert len(latest_policy_arns) == 0
+
+        # Test the code paths that synchronize the attached policies for a user
+        policy_arns = [
+            "arn:aws:iam::aws:policy/AmazonS3ReadOnlyAccess",
+        ]
+        updates = {
```

Comment:
Can you please add logic to this test to cover the case of changing the `UserName` via the custom update code path?

Reply:
This is a tricky one, `User.Name` is the unique identifier the controllers uses to describe/update/delete a user. Changing this field will cause the controller create a new user instead of updating the existing one. Shall we mark `User.Name` as immutable field instead?
related to https://github.com/aws-controllers-k8s/community/issues/1111